File: squares-of-a-sorted-array.cpp
/*
 *Solution for problem 977
 *https://leetcode.com/problems/squares-of-a-sorted-array/
 */

class Solution {
public:
    vector<int> sortedSquares(vector<int>& A) {
        for(int i=0; i<A.size(); i++)
            A[i] = A[i] * A[i];
        sort(A.begin(), A.end());
        return A;
    }
};


/*Big-O(n) Solution using two pointers
 *
 *
 *class Solution {
public:
    vector<int> sortedSquares(vector<int>& A) {
        int N = A.size();
        int i, j=0;
        while( j<N && A[j] < 0)
            j += 1;
        i = j - 1;
        vector<int> ans;
        while( i>=0 && j<N ){
            if(A[i] * A[i] < A[j] * A[j]){
                ans.push_back(A[i]*A[i]);
                i -= 1;
            }
            else{
                ans.push_back(A[j] * A[j]);
                j += 1;
            }
        }
        while( i>=0 ){
            ans.push_back(A[i] *A[i]);
            i -= 1;
        }
        while( j<N ){
            ans.push_back(A[j] * A[j]);
            j += 1;
        }
        return ans;
    }
};
 *
 */

File: solutions/squares-of-a-sorted-array.py
"""
977. Squares of a Sorted Array
https://leetcode.com/problems/squares-of-a-sorted-array/
"""

# Solution 1:
# Time Complexity: O(N log N)
class Solution:
    def sortedSquares(self, A: List[int]) -> List[int]:
        return sorted[x * x for x in A]





# Solution 2:
# Time ComplexityL O(N)
class Solution:
    def sortedSquares(self, A: List[int]) -> List[int]:
        LenA = len(A)
        j = 0

        # j will hold the first positive number's index
        while j < LenA and A[j] < 0:
            j += 1

        # i will hold the highest negative number
        i = j - 1

        # output list to hold resultant array
        output = []

        while i >= 0 and j < LenA:
            if A[i]**2 < A[j]**2:
                output.append(A[i]**2)
                i -= 1
            else:
                output.append(A[j]**2)
                j += 1

        # if there are still some element left unvisited
        while i >= 0:
            output.append(A[i]**2)
            i -= 1
        while j < LenA:
            output.append(A[j]**2)
            j += 1

        # return output list
        return output

